Add documentation for using ansible-test with collections (#66041)
ansible/ansible: docs/docsite/rst/dev_guide/developing_collections.rst

------------------

Add a 'per plugin type' specific subdirectory here, including ``module_utils`` which is usable not only by modules, but by most plugins by using their FQCN. This is a way to distribute modules, lookups, filters, and so on, without having to import a role in every play.

> 

Reviewer: Is this necessary?

Author: I didn't see the dash in the latest version. Please verify I'm not missing anything.

Reviewer: Dash? It's a paragraph separator removal, right?

Reviewer 2: This change is removing an empty line between the two paragraphs.

Add a 'per plugin type' specific subdirectory here, including ``module_utils`` which is usable not only by modules, but by most plugins by using their FQCN. This is a way to distribute modules, lookups, filters, and so on without having to import a role in every play.

=======
Add a 'per plugin type' specific subdirectory here, including ``module_utils`` which is usable not only by modules, but by most plugins by using their FQCN. This is a way to distribute modules, lookups, filters, and so on, without having to import a role in every play.

> >>>>>>> 748ef10d7e... Update developing_collections.rst

Reviewer 2: There are some rebase artefacts left in this PR. Could you please remove them?

tests directory
----------------

> Executing sanity, integration, and unit tests is largely identical to performing these tests within Ansible's mainline code. ``ansible-test`` executes against modules implemented from the relative path ``ansible_collections/..`` For example, ``ansible-test`` may be executed against IOS modules from the default collection location of ``~/.ansible/collections/ansible_collections/cisco/ios/``. Tests may also be executed while in a custom directory, such as ``~/work/ansible_collections/cisco/ios/``, as long as it is the current path. Also, collections must not be executed from within Ansible's source tree and cannot reside within a directory comprised of a symbolic link.

Reviewer: This is a very long line. Mind multi-lining it to follow the style of this doc?

Author: Updated.

tests directory
----------------

> Executing sanity, integration, and unit tests is largely identical to performing these tests within Ansible's mainline code. ``ansible-test`` executes against modules implemented from the relative path ``collections//.. For example, ``ansible-test`` may be executed against IOS modules from the default collection location of ``~/.ansible/collections/cisco/ios/``. Tests may also be executed while in a custom directory, such as ``~/work/ansible/collections/cisco/ios/``, as long as it is the current path.

Reviewer 3: ```suggestion
Executing sanity, integration, and unit tests is largely identical to performing these tests within Ansible's mainline code. ``ansible-test`` executes against modules implemented from the relative path ``collections/.. For example, ``ansible-test`` may be executed against IOS modules from the default collection location of ``~/.ansible/collections/cisco/ios/``. Tests may also be executed while in a custom directory, such as ``~/work/ansible/collections/cisco/ios/``, as long as it is the current path.
```

> Executing sanity, integration, and unit tests is largely identical to performing these tests within Ansible's mainline code. ``ansible-test``'s targets must execute from a directory path with the following format: ``collections/<namespace>/<collection>``. For example, ``ansible-test`` may be executed against IOS modules from the default collection location of ``~/.ansible/collections/cisco/ios/``. Tests may also be executed while in a custom directory, such as ``~/work/ansible/collections/cisco/ios/``, as long as it is the current path.

Reviewer 4: Recommend re-phrasing the second sentence here for clarity: ``ansbile-test`` executes against modules implemented from the relative path ``collections/<namespace>/<collections>.

Author: I updated. What do you think now?

Sanity tests are executed using ``ansible-test sanity`` syntax. Refer to :ref:`_testing_sanity` for details.

> Integration tests are executed using ``ansible-test integration`` syntax. Network integration tests are executed using ``ansible-test network-integration`` syntax. Refer to :ref:`_testing_integration` for details.

Reviewer 5: ```suggestion
Integration tests are executed using the ``ansible-test integration`` command. Network integration tests are executed using the ``ansible-test network-integration`` command. Windows integration tests use the ``ansible-test windows-integration`` command. Refer to :ref:`_testing_integration` for details.
```

.. note::

>     The collection name must be in ``tests/integration/target-prefixes.network`` for network-integration tests to recognize the targets.

Reviewer 5: ```suggestion
    For ``ansible-test`` to recognize tests as ``network-integration`` tests, the test targets must have a ``network`` entry in their ``aliases`` file. Alternatively, if the modules and targets share a common prefix, such as ``ios_`` then that prefix (without the underscore) can be added to `tests/integration/target-prefixes.network``, with one prefix listed per line.
    
    For ``ansible-test`` to recognize tests as ``windows-integration`` tests, the test targets must have a ``windows`` entry in their ``aliases`` file. Alternatively, the modules and targets may use the ``win_`` prefix.
```

tests directory
----------------

> Executing sanity, integration, and unit tests is largely identical to performing these tests within Ansible's mainline code. ``ansible-test`` executes against modules implemented from the relative path ``collections/.. For example, ``ansible-test`` may be executed against IOS modules from the default collection location of ``~/.ansible/collections/cisco/ios/``. Tests may also be executed while in a custom directory, such as ``~/work/ansible/collections/cisco/ios/``, as long as it is the current path.

Reviewer 6: ```suggestion
Executing sanity, integration, and unit tests are largely identical to performing these tests within Ansible's mainline code. ``ansible-test`` executes against modules implemented from the relative path ``collections/..``. For example, ``ansible-test`` may be executed against IOS modules from the default collection location of ``~/.ansible/collections/cisco/ios/``. Tests may also be executed while in a custom directory, such as ``~/work/ansible/collections/cisco/ios/``, as long as it is the current path.
```

> Executing sanity, integration, and unit tests is largely identical to performing these tests within Ansible's mainline code. ``ansible-test`` executes against modules implemented from the relative path ``collections/.. For example, ``ansible-test`` may be executed against IOS modules from the default collection location of ``~/.ansible/collections/cisco/ios/``. Tests may also be executed while in a custom directory, such as ``~/work/ansible/collections/cisco/ios/``, as long as it is the current path.

Reviewer 5: The important directory is `ansible_collections` rather than `collections`:

```suggestion
Executing sanity, integration, and unit tests is largely identical to performing these tests within Ansible's mainline code. ``ansible-test`` executes against modules implemented from the relative path ``ansible_collections/..`` For example, ``ansible-test`` may be executed against IOS modules from the default collection location of ``~/.ansible/collections/ansible_collections/cisco/ios/``. Tests may also be executed while in a custom directory, such as ``~/work/ansible_collections/cisco/ios/``, as long as it is the current path.
```

Two other things should also be explained:

1. The `ansible_collections` directory must not be within the Ansible source tree. If it is, `ansible-test` will try to test the Ansible source instead of the collection.
2. Symbolic links cannot be used to construct the required path. The real path is seen by Ansible and other tools.

Sanity tests are executed using ``ansible-test sanity`` syntax. Refer to :ref:`_testing_sanity` for details.

> Integration tests are executed using the ``ansible-test integration`` command. Network integration tests are executed using the ``ansible-test network-integration`` command. Windows integration tests use the ``ansible-test windows-integration`` command. Refer to :ref:`_testing_integration` for details.

Reviewer 6: ```suggestion
* Integration tests are executed using the ``ansible-test integration`` command.

* Network integration tests are executed using the ``ansible-test network-integration`` command.

* Windows integration tests use the ``ansible-test windows-integration`` command.

Refer to :ref:`_testing_integration` for details.
```

.. note::

>     For ``ansible-test`` to recognize tests as ``network-integration`` tests, the test targets must have a ``network`` entry in their ``aliases`` file. Alternatively, if the modules and targets share a common prefix, such as ``ios_`` then that prefix (without the underscore) can be added to `tests/integration/target-prefixes.network``, with one prefix listed per line.

Reviewer 6: ```suggestion
    For ``ansible-test`` to recognize tests as ``network-integration`` tests, the test targets must have a ``network`` entry in their ``aliases`` file. Alternatively, if the modules and targets share a common prefix, such as ``ios_`` then that prefix (without the underscore) can be added to ``tests/integration/target-prefixes.network``, with one prefix listed per line.
```

>     For ``ansible-test`` to recognize tests as ``windows-integration`` tests, the test targets must have a ``windows`` entry in their ``aliases`` file. Alternatively, the modules and targets may use the ``win_`` prefix.

Reviewer 6: ```suggestion
    For ``ansible-test`` to recognize tests as ``windows-integration`` tests, the test targets must have a ``windows`` entry in their ``aliases`` file. Alternatively, the modules and targets may use the ``win_`` prefix.
```

> Unit tests are executed using ``ansible-test unit`` syntax. Refer to :ref:`_testing_units` for details.

Reviewer 5: ```suggestion
Unit tests are executed using ``ansible-test units`` syntax. Refer to :ref:`_testing_units` for details.
```

tests directory
----------------

> Executing sanity, integration, and unit tests is largely identical to performing these tests within Ansible's mainline code. 

Reviewer 2: There is already a whole section in the document on running tests in collections: https://docs.ansible.com/ansible/devel/dev_guide/developing_collections.html#testing-collections

I think this should be integrated with that section, instead of repeating the information at another place. (Two paragraphs after your new addition, there's already a link to that section.)
